### PR TITLE
API doc updates

### PR DESCRIPTION
**Pull request by alexwishart**
Part of work for this [issue](https://github.ibm.com/arf/planning-swift-solutions/issues/493). @helenmasters can't add you as a reviewer, so tagging you instead 🙂

### CODE REVIEW

**Comment by helenmasters** [comment]
Alex, as I can't add myself as a reviewer I'll put some comments here.

- I think it'd be useful to keep the basic sample usage in the README.md to minimise the number of clicks that I would need to perform as a user before finding the information I was looking for.

- In the Statistics struct I would like to see more clarity on the descriptions for hits, misses and numberOfKeys, something more like -

hits - The total number of times an access attempt successfully retrieved an item from the cache.
misses - The total number of times an access attempt did not find an item in the cache.
numberOfKeys - The number of entries currently in the cache. (I realise that this is called numberOfKeys but that is not what a user would be thinking about when using this I don't think...).

- Personally I don't like the description "The statistics of the cache." but something like "Statistics about the cache, including the number of hits and misses and the number of items in the cache." is maybe too wordy? I'm not sure please can you take another look at this.

**Comment by helenmasters** [comment]
In the top level description the KituraCache class description I think the 'object with a Hashable key' needs some work (and the Hashable needs to be in quotes) as it's not clear if the key needs to be part of the object or not. The second sentence "Time To Live can be specified per entry, or will take a default value specified when the cache is initialised." I found confusing, maybe something like the following would be clearer

"The Time to live (TTL), i.e. the lifespan of the data, can be specified either when the cache is initialised or when an object is added."

**Comment by helenmasters** [comment]
On the init method the description for the defaultTTL parameter doesn't make sense to me.

"The default Time to Live value (in seconds). Used for cache entries for which the TTL is not specified otherwise. If it is not specified here, the default value is 0 which maps to infinity."

I think that the second two sentences need work - we certainly don't need to talk about infinity! Something like "...the default TTL value is 0 which means that objects in the cache will not expire." might help.

**Comment by alexwishart** [comment]
@helenmasters Have made the requested changes, plus a few others that I managed to find this afternoon. Have also added new badges and logo to the README - what do you think? 🙂 Still haven't found a nice way of linking to the API docs apart from in the badge, will chat to you tomorrow about how we can fix this.

**Comment by rolivieri** [comment]
Just added @helenmasters as the reviewer.

File: README.md [hunk 1]
```diff
@@ -1,75 +1,74 @@
...
-- *key/s* have to be *Hashable*
 
-### Set TTL of an entry in the cache
+**To retrieve an entry from the cache:**
 ```swift
-public func setTTL<T: Hashable>(_ ttl: UInt, forKey key: T) -> Bool
+let cache = KituraCache()
+...
+if let item = cache.object(forKey: 1) {
+//Object with key of 1 retrieved from cache.
```
Comment (helenmasters) [inline README.md]: It might be clearer to indent both the comment and ... which are within the else { }

File: README.md [hunk 2]
```diff
@@ -1,75 +1,74 @@
...
+...
+}
+else {
+//No object stored in cache with key of 1.
+...
```
Comment (helenmasters) [inline README.md]: Same comment as above regarding indentation.

File: README.md [hunk 3]
```diff
@@ -1,75 +1,74 @@
...
-### Retrieve cache statistics
-Cache statistics are stored in
-```swift
-public private(set) var statistics: Statistics
-```
-**Statistics struct contains:**
-   - *hits* the number of cache hits
-   - *misses* the number of cache misses
-   - *numberOfKeys* the total number of keys in the cache
+_Refer to [KituraCache](https://ibm-swift.github.io/Kitura-Cache/Classes/KituraCache) and [Statistics](https://ibm-swift.github.io/Kitura-Cache/Structs/Statistics) for more information and further configuration._
```
Comment (helenmasters) [inline README.md]: I don't think the words "and further configuration" add anything useful to this sentence - please can you remove them.

File: Sources/KituraCache/Statistics.swift [hunk 4]
```diff
@@ -16,16 +16,37 @@
...
+    ````swift
+    let cache = KituraCache()
+    ...
+    let numMisses = cache.statistics.misses
+    ```
+    */
     public internal(set) var misses = 0
     
-    /// The total number of keys in the cache.
+    /** The total number of entries curently in the cache.
```
Comment (helenmasters) [inline Sources/KituraCache/Statistics.swift]: curently - spelling needs fixing to currently!

File: Sources/KituraCache/Statistics.swift [hunk 5]
```diff
@@ -16,16 +16,37 @@
...
+     ### Usage Example: ###
```
Comment (helenmasters) [inline Sources/KituraCache/Statistics.swift]: This is displaying as "The total number of entries curently in the cache. ### Usage Example: ###" when it's loaded so needs fixing.

File: Sources/KituraCache/Statistics.swift [hunk 6]
```diff
@@ -16,16 +16,37 @@
...
+    ````swift
+     let cache = KituraCache()
+     ...
+     let numHits = cache.statistics.hits
+     ```
+    */
     public internal(set) var hits = 0
     
-    /// The number of the cache misses.
+    /** The total number of times an access attempt was unable to retrieve an entry in the cache.
```
Comment (helenmasters) [inline Sources/KituraCache/Statistics.swift]: I think "from the cache" would be better than "in the cache" here.

File: Sources/KituraCache/KituraCache.swift [hunk 7]
```diff
@@ -64,12 +69,22 @@ public class KituraCache {
...
-    /// - Parameter object: The data object.
-    /// - Parameter forKey: The key for the data.
-    /// - Parameter withTTL: The optional Time to Live value in seconds for the entry. If not specified,
-    ///                     the default TTL is used.
+    //MARK: Adding objects
+    
+    /**
+     Adds a new entry or updates the existing entry if the key is already associated with an object in the cache. The lifespan of the entry (in seconds) in the cache can be set using the optional withTTL parameter.
+     ### Usage Example: ###
+     In this example, item is an instance of a struct object with an id field which conforms to Hashable.
```
Comment (helenmasters) [inline Sources/KituraCache/KituraCache.swift]: struct and Hashable should be quoted.

File: Sources/KituraCache/KituraCache.swift [hunk 8]
```diff
@@ -89,11 +104,28 @@ public class KituraCache {
...
-    ///
-    /// - Parameter forKey: The key of the entry to retrieve.
-    /// - Returns: The object stored in the cache for the key.
-    /// - Note: The return value will be nil if there is no object in the cache with the specified key.
+    //MARK: Retrieving objects
+    
+    /**
+     Retrieve an object from the cache for a specified key.
+     ### Usage Example: ###
+     In this example, item has been stored in the cache with an integer key
```
Comment (helenmasters) [inline Sources/KituraCache/KituraCache.swift]: Full stop at end.

File: Sources/KituraCache/KituraCache.swift [hunk 9]
```diff
@@ -102,23 +134,75 @@ public class KituraCache {
...
         removeObjects(forKeys: [key])
     }
     
-    /// Remove objects from the cache.
-    ///
-    /// - Parameter forKeys: The keys of the entries to remove.
+    /**
+     Remove objects from the cache for multiple, specified keys.
+     ### Usage Example: ###
+     In this example, objects have been stored in the cache with an integer key
```
Comment (helenmasters) [inline Sources/KituraCache/KituraCache.swift]: Full stop at end.

File: Sources/KituraCache/KituraCache.swift [hunk 10]
```diff
@@ -102,23 +134,75 @@ public class KituraCache {
...
+        }
+        return keys
+    }
+    
+    //MARK: Removing objects
+    
+    /**
+     Remove an object from the cache for a specified key.
+     ### Usage Example: ###
+     In this example, objects have been stored in the cache with an integer key
```
Comment (helenmasters) [inline Sources/KituraCache/KituraCache.swift]: Full stop at end.

File: Sources/KituraCache/KituraCache.swift [hunk 11]
```diff
@@ -102,23 +134,75 @@ public class KituraCache {
...
         removeObjects(forKeys: keys)
     }
     
-    /// Remove objects from the cache.
-    ///
-    /// - Parameter forKeys: An array of the keys of the entries to remove.
+    /**
+     Remove objects from the cache for multiple, specified keys provided in an array.
+     ### Usage Example: ###
+     In this example, objects have been stored in the cache with an integer key
```
Comment (helenmasters) [inline Sources/KituraCache/KituraCache.swift]: Full stop at end.

File: Sources/KituraCache/KituraCache.swift [hunk 12]
```diff
@@ -145,11 +237,21 @@ public class KituraCache {
...
-    ///
-    /// - Parameter ttl: The Time to Live value in seconds.
-    /// - Parameter forKey: The key of the entry to set its TTL.
-    /// - Returns: True if the TTL was successfully set, and false if the key doesn't exist.
+    //MARK: Changing TTL for an entry
+    
+    /**
+     Set the Time to Live value (in seconds) for a cache entry.
+     ### Usage Example: ###
+     In this example, objects have been stored in the cache with an integer key
```
Comment (helenmasters) [inline Sources/KituraCache/KituraCache.swift]: Full stop at end.

**Review by helenmasters — CHANGES_REQUESTED** [review]

**Comment by alexwishart** [comment]
@helenmasters have made the requested fixes 🙂

File: Sources/KituraCache/Statistics.swift [hunk 13]
```diff
@@ -39,12 +41,13 @@ public struct Statistics {
...
     
-    /** The total number of entries curently in the cache.
+    /**
+     The total number of entries currently in the cache.
      ### Usage Example: ###
      ````swift
      let cache = KituraCache()
      ...
-     let numKeys = cache.statistics.numberOfKeys
+     let numMisses = cache.statistics.numberOfKeys
```
Comment (helenmasters) [inline Sources/KituraCache/Statistics.swift]: This line looks like it was updated in error?

File: Sources/KituraCache/Statistics.swift [hunk 14]
```diff
@@ -29,7 +30,8 @@ public struct Statistics {
     */
     public internal(set) var hits = 0
     
-    /** The total number of times an access attempt was unable to retrieve an entry in the cache.
+    /**
+     The total number of times an access attempt was unable to retrieve an entry in the cache.
```
Comment (helenmasters) [inline Sources/KituraCache/Statistics.swift]: Please can you change "in the cache" to "from the cache"

File: Sources/KituraCache/Statistics.swift [hunk 15]
```diff
@@ -19,7 +19,8 @@
 /// Statistics about the cache.
 public struct Statistics {
     
-    /** The total number of times an access attempt successfully retrieved an entry from the cache.
+    /**
+     The total number of times an access attempt successfully retrieved an entry in the cache.
```
Comment (helenmasters) [inline Sources/KituraCache/Statistics.swift]: Please can you change "in the cache" to "from the cache"

File: Sources/KituraCache/KituraCache.swift [hunk 16]
```diff
@@ -19,26 +19,31 @@ import Dispatch
...
-    ///                         the TTL was not specified otherwise.
-    /// - Parameter checkFrequency: The frequency (in seconds) to check for expired entries.
-    public init(defaultTTL: UInt = 0, checkFrequency: UInt = 600) {
+    /**
+     Initialize an instance of KituraCache.
+     ### Usage Example: ###
+     ````swift
+     let cache = KituraCache(defaultTTL: 3600, checkFrequency: 600)
+     ````
+     - Parameter defaultTTL: The Time to Live value (in seconds) of an entry if a TTL value is not specified when the entry is later added. If the default TTL is not defined during this initialisation, entries in the cache are configured to not expire.
```
Comment (helenmasters) [inline Sources/KituraCache/KituraCache.swift]: the entry is later added -> "an entry is added later" I think would be clearer.

**Review by helenmasters — APPROVED** [review]
The changes all look good to me.  
@rolivieri can you do a technical review please, then if you're happy we'll be good to go.

Also @rolivieri do you think us linking to an icon in the Kitura central repository for the README banner headers will cause us issues (rather than having a copy in each local repo)?

**Comment by helenmasters** [comment]
@tunniclm I can't add you as a reviewer, so tagging you instead 🙂, please can you do a technical review so that this can be merged.

**Review by tunniclm — APPROVED** [review]
